leakpro/attacks/gia_attacks/invertinggradients.py

```diff
@@ -0,0 +1,167 @@
+"""Geiping, Jonas, et al. "Inverting gradients-how easy is it to break privacy in federated learning?."."""
+import torch
+from torch.utils.data import DataLoader, TensorDataset
+
+from leakpro.attacks.gia_attacks.abstract_gia import AbstractGIA
+from leakpro.import_helper import Callable, Self
+from leakpro.metrics.attack_result import GIAResults
+from leakpro.user_inputs.abstract_gia_input_handler import AbstractGIAInputHandler
+
+
+class InvertingGradients(AbstractGIA):
+    """Gradient inversion attack by Geiping et al."""
+
+    def __init__(self: Self, handler: AbstractGIAInputHandler, configs: dict) -> None:
+        super().__init__(handler)
+        self.handler = handler
+        self.logger.info("Inverting gradient initialized :)")
+        self.t_v_scale = configs.get("total_variation")
+        self.attack_lr = configs.get("attack_lr")
+
+    def description(self:Self) -> dict:
+        """Return a description of the attack."""
+        title_str = "Inverting gradients"
+        reference_str = "Geiping, Jonas, et al. Inverting gradients-how easy is it to break privacy in federated learning?(2020)."
+        summary_str = ""
+        detailed_str = ""
+        return {
+            "title_str": title_str,
+            "reference": reference_str,
+            "summary": summary_str,
+            "detailed": detailed_str,
+        }
+
+    def prepare_attack(self:Self) -> None:
+        """Prepare the attack.
+
+        Args:
+        ----
+            self (Self): The instance of the class.
+
+        Returns:
+        -------
+            None
+
+        """
+        self.handler.target_model.eval()
+        self.client_loader = self.handler.client_data
+        self.data_mean, self.data_std = self.handler.get_meanstd()
+        self.reconstruction = self.handler.init_at_image()
+
+        # Extract the labels from self.handler.client_data
+        labels = []
+        for _, label in self.handler.client_data:
+            labels.extend(label.numpy())
+        labels = torch.tensor(labels)
+        reconstruction_dataset = TensorDataset(self.reconstruction, labels)
+        self.reconstruction_loader = DataLoader(reconstruction_dataset, batch_size=32, shuffle=True)
+
+        client_gradient = self.handler.train(self.client_loader, self.handler.get_optimizer())
+        self.client_gradient = [p.detach() for p in client_gradient]
+
+    def run_attack(self:Self) -> GIAResults:
+        """Run the attack and return the combined metric result.
+
+        Returns
+        -------
+            GIAResults: Container for results on GIA attacks.
+
+        """
+        self.reconstruction.requires_grad = True
+        optimizer = torch.optim.Adam([self.reconstruction], lr=self.attack_lr)
+        scheduler = torch.optim.lr_scheduler.MultiStepLR(optimizer,
```

Comment:
Is this required?

Reply:
yes, torch optimizers expects something iterable thats why we put it in a list.